{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copyright 2019 Google LLC\n",
    "# Copyright 2025 Anonymized Authors\n",
    "\n",
    "# Licensed under the Apache License, Version 2.0 (the \"License\"); \n",
    "# you may not use this file except in compliance with the License. \n",
    "# You may obtain a copy of the License at\n",
    "# https://www.apache.org/licenses/LICENSE-2.0\n",
    "\"\"\"\n",
    "This notebook benchmarks GENE and regularized evolution on NAS-Bench-101.\n",
    "It uses a population size of 50 and mutation rate of 0.72 for RE. \n",
    "\n",
    "Requirements: \n",
    "\n",
    "-This notebook requires that  tensorflow and numpy be installed within the \n",
    "Python environment you are running this script in. \n",
    "\n",
    "\"\"\"\n",
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import numpy as np\n",
    "import random\n",
    "\n",
    "import sys; sys.path.append('..')\n",
    "from utils.nas_utils import load_nasbench,  NASBenchConstants, NASBench, SpecOneHot, run_revolution_search, init_pop\n",
    "from utils.plotting import plot_all\n",
    "from IPython.display import clear_output\n",
    "from typing import Callable, List, Any\n",
    "\n",
    "nasb = load_nasbench()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we define our GENE algorithm: \n",
    "\n",
    "def greedy_selection(parents):\n",
    "    \"\"\"Selects greedy n parent pairs\"\"\"\n",
    "    p_size = len(parents)\n",
    "    population = np.asarray([p[0] for p in parents])\n",
    "\n",
    "    summed_population = []\n",
    "    for p1 in range(p_size):\n",
    "        for p2 in range(p_size):\n",
    "            if p1 < p2: # only upper triangle\n",
    "                sum = population[p1] + population[p2]\n",
    "            else:\n",
    "                sum = 0\n",
    "            summed_population.append(sum)\n",
    "\n",
    "    samples = np.argsort(summed_population)[-p_size:]\n",
    "    row_indices, col_indices = np.meshgrid(np.arange(p_size),np.arange(p_size))\n",
    "    mapping = np.reshape(np.stack([col_indices,row_indices],axis=-1), [-1,2])\n",
    "    pairings = mapping[samples]\n",
    "\n",
    "    paired_parents = []\n",
    "    for i in range(len(pairings)):\n",
    "        p1 = parents[pairings[i,0]]\n",
    "        p2 = parents[pairings[i,1]]\n",
    "        paired_parents.append((p1,p2))\n",
    "\n",
    "    return paired_parents\n",
    "\n",
    "\n",
    "def crossover(nasb, parents):\n",
    "    \"\"\"Performs vertical crossover on parent pairs, ensuring valid offspring.\"\"\"\n",
    "    \n",
    "    def cross(p1, p2):\n",
    "        if p1.original_matrix.shape != p2.original_matrix.shape:\n",
    "            print(\"Shape mismatch\")\n",
    "            return p1\n",
    "\n",
    "        for _ in range(5):\n",
    "            i = random.randint(1, NASBenchConstants.NUM_VERTICES - 1)\n",
    "            m = np.copy(p1.original_matrix)\n",
    "            m[:i, :] = p2.original_matrix[:i, :]\n",
    "            o = p1.original_ops[:i] + p2.original_ops[i:]\n",
    "            new_spec = SpecOneHot(m, o)\n",
    "            if nasb.is_valid(new_spec):\n",
    "                return new_spec\n",
    "        return p1\n",
    "\n",
    "    return [cross(p[0][1], p[1][1]) for p in parents]\n",
    "\n",
    "\n",
    "# guided mutation for nasbench specific encoding\n",
    "def guided_mutation(nasb, population, variant):\n",
    "    psize = len(population)\n",
    "    \n",
    "    inputs = [x.to_flat() for x in population]\n",
    "    mutation_inputs = np.array(inputs)\n",
    "    summed = np.sum(mutation_inputs, axis=0)\n",
    "\n",
    "    # Define the probabilities for mutation\n",
    "    probs_mat_1 = summed[:SpecOneHot.flat_matrix:2] / psize \n",
    "    probs_mat_1 /= np.sum(probs_mat_1)  \n",
    "    probs_mat_0 = 1 - probs_mat_1  \n",
    "    probs_mat_0 /= np.sum(probs_mat_0)  \n",
    "\n",
    "    probs_ops_1 = summed[SpecOneHot.flat_matrix:] / psize\n",
    "    probs_ops_1 /= np.sum(probs_ops_1)  \n",
    "    probs_ops_0 = 1 - probs_ops_1\n",
    "    probs_ops_0 /= np.sum(probs_ops_0)  \n",
    "\n",
    "    if variant==0:\n",
    "        probs_ops = probs_ops_0\n",
    "    else:\n",
    "        probs_ops = probs_ops_1\n",
    "\n",
    "    children = []\n",
    "    for i in range(psize):\n",
    "        while True:\n",
    "            child_spec = mutation_inputs[i, :].astype(int)\n",
    "\n",
    "            # Select indices based on the probabilities using np.random.choice (similar to torch.multinomial)\n",
    "            mat_index_1 = np.random.choice(len(probs_mat_0), 1, p=probs_mat_0)[0]\n",
    "            mat_index_0 = np.random.choice(len(probs_mat_1), 1, p=probs_mat_1)[0]\n",
    "\n",
    "            ops_index = np.random.choice(len(probs_ops), 1, p=probs_ops)[0] + SpecOneHot.flat_matrix\n",
    "\n",
    "            # Modify the child_spec array based on the selected indices\n",
    "            base = (ops_index - (ops_index % 3))  # Ensure base is aligned to 3 operations\n",
    "            remove = [base, base + 1, base + 2]\n",
    "            child_spec[remove] = 0\n",
    "            child_spec[ops_index] = 1\n",
    "\n",
    "            # mutate adjacency matrix\n",
    "            child_spec[mat_index_1*2] = 1\n",
    "            child_spec[mat_index_1*2+1] = 0\n",
    "\n",
    "            child_spec[mat_index_0*2+1] = 1\n",
    "            child_spec[mat_index_0*2] = 0\n",
    "\n",
    "            spec = SpecOneHot.spec_from_flat(child_spec)\n",
    "            \n",
    "            if nasb.is_valid(spec):\n",
    "                children.append(spec)\n",
    "                break\n",
    "\n",
    "    return children\n",
    "\n",
    "\n",
    "def run_xevolution(\n",
    "    nasbench: NASBench,\n",
    "    selection_algorithm: Callable[[List], Any],\n",
    "    crossover_algorithm: Callable[[NASBench, List], Any],\n",
    "    mutation_algorithm: Callable[[NASBench, List], Any],\n",
    "    max_time_budget=5e6,\n",
    "    p_size=50,    \n",
    "):\n",
    "    \"\"\"Run a single roll-out of evolution to a fixed time budget.\"\"\"\n",
    "    population, times, best_valids, best_tests = init_pop(nasbench, max_time_budget, p_size)\n",
    "    while True:\n",
    "        selected_candidates = selection_algorithm(population)  \n",
    "        crossed_parents = crossover_algorithm(nasbench, selected_candidates)\n",
    "        children = mutation_algorithm(nasbench, crossed_parents)\n",
    "        population = []\n",
    "        for child in children: \n",
    "            data = nasbench.query(child)\n",
    "            time_spent, _ = nasbench.get_budget_counters()\n",
    "            times.append(time_spent)\n",
    "            population.append((data[\"validation_accuracy\"], child))\n",
    "            if data[\"validation_accuracy\"] > best_valids[-1]:\n",
    "                best_valids.append(data[\"validation_accuracy\"])\n",
    "                best_tests.append(data[\"test_accuracy\"])\n",
    "            else:\n",
    "                best_valids.append(best_valids[-1])\n",
    "                best_tests.append(best_tests[-1])\n",
    "\n",
    "        if time_spent> max_time_budget:\n",
    "            break\n",
    "\n",
    "    return times, best_valids, best_tests\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# comparing GENE-0, GENE-1 and regularized evolution\n",
    "gm0 = lambda nasb, population : guided_mutation(nasb, population, 0)\n",
    "gm1 = lambda nasb, population : guided_mutation(nasb, population, 1)\n",
    "\n",
    "exp1 = {\n",
    "    \"data\": \n",
    "    {\n",
    "        \"GENE-0\" : [[],\"Dark Blue\"],\n",
    "        \"GENE-1\" : [[],\"Red Orange\"],\n",
    "        \"RE\" : [[],\"Very Dark Gray\"],\n",
    "\n",
    "    },\n",
    "    \"config\": \n",
    "    {\n",
    "        \"budget\" : int(1e7),\n",
    "        \"limits\" : (0.9395, 0.9435),\n",
    "        \"n\" : 1000,\n",
    "        \"print_every\" : 2,\n",
    "        \"confidence_intervall\" : True,\n",
    "        \"pvalue\" : 0.05,\n",
    "        \"significant_areas\": False,\n",
    "        \"dataset\" : \"test\",\n",
    "   }\n",
    "}\n",
    "\n",
    "budget = exp1[\"config\"][\"budget\"]\n",
    "for run in range(exp1[\"config\"][\"n\"]):\n",
    "    nasb.reset_budget_counters()\n",
    "    times, best_valid, best_test = run_xevolution(nasb, greedy_selection, crossover, gm0, budget, 50)\n",
    "    exp1[\"data\"][\"GENE-0\"][0].append((times, best_valid, best_test))\n",
    "\n",
    "    nasb.reset_budget_counters()\n",
    "    times, best_valid, best_test = run_xevolution(nasb, greedy_selection, crossover, gm1, budget, 50)\n",
    "    exp1[\"data\"][\"GENE-1\"][0].append((times, best_valid, best_test))\n",
    "\n",
    "    nasb.reset_budget_counters()\n",
    "    times, best_valid, best_test = run_revolution_search(nasb, budget, 50,10,0.72)\n",
    "    exp1[\"data\"][\"RE\"][0].append((times, best_valid, best_test))\n",
    "\n",
    "\n",
    "    if (run % exp1[\"config\"][\"print_every\"] == 0):\n",
    "        clear_output(wait=True)\n",
    "        fig, ax = plt.subplots()\n",
    "        plot_all(exp1, ax)\n",
    "        plt.show()\n",
    "        print('Running repeat %d' % (run + 1))\n",
    "\n",
    "clear_output(wait=True)\n",
    "fig, ax = plt.subplots()\n",
    "means1 , _ =  plot_all(exp1, ax)\n",
    "plt.savefig('GENE_benchmark_all.png', dpi=600)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save results to folder\n",
    "! this could overwrite existing files\n",
    "\n",
    "date = \"3_17\"\n",
    "f1 = os.path.join(\"..\",\"generated\",\"models\", \"GENE\", f\"GENE-0_1e7_S_1000_{date}.json\")\n",
    "f2 = os.path.join(\"..\",\"generated\",\"models\", \"GENE\", f\"GENE-1_1e7_S_1000_{date}.json\")\n",
    "f3 = os.path.join(\"..\",\"generated\",\"models\", \"GENE\", f\"RE_1e7_S1000_{date}.json\")\n",
    "\n",
    "with open(f1, 'w') as file:\n",
    "    json.dump(exp1[\"data\"][\"GENE-0\"][0], file)\n",
    "\n",
    "with open(f2, 'w') as file:\n",
    "    json.dump(exp1[\"data\"][\"GENE-1\"][0], file)\n",
    "\n",
    "with open(f3, 'w') as file:\n",
    "    json.dump(exp1[\"data\"][\"RE\"][0], file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# only print specific plots from import\n",
    "\n",
    "exp1a = {\n",
    "    \"data\": \n",
    "    {\n",
    "        # \"GENE-0\" : [[],\"Dark Blue\"], # comment which one to plot\n",
    "        \"GENE-1\" : [[],\"Dark Blue\"],\n",
    "        \"RE\" : [[],\"Very Dark Gray\"],\n",
    "\n",
    "    },\n",
    "    \"config\": \n",
    "    {\n",
    "        \"budget\" : int(1e7),\n",
    "        \"limits\" : (0.9395, 0.9435),\n",
    "        \"n\" : 1000,\n",
    "        \"print_every\" : 2,\n",
    "        \"confidence_intervall\" : True,\n",
    "        \"pvalue\" : 0.05,\n",
    "        \"significant_areas\": False,\n",
    "        \"dataset\" : \"test\",\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "f3 = os.path.join(\"..\",\"generated\",\"models\", \"GENE\", \"new_gm1_1e7.json\")\n",
    "f4 = os.path.join(\"..\",\"generated\",\"models\", \"GENE\", \"new_re_1e7.json\")\n",
    "\n",
    "with open(f3, 'r') as file:\n",
    "    exp1a[\"data\"][\"GENE-1\"][0] = json.load(file)\n",
    "\n",
    "with open(f4, 'r') as file:\n",
    "    exp1a[\"data\"][\"RE\"][0] = json.load(file)\n",
    "    \n",
    "print(\"importing done.\")\n",
    "fig, ax = plt.subplots()\n",
    "means1a , _ =  plot_all(exp1a, ax)\n",
    "# plt.savefig('GENE-1_1e7_S_1000.png', dpi=600)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# only print specific plots from previous experiment\n",
    "\n",
    "exp1b = {\n",
    "    \"data\": \n",
    "    {\n",
    "        \"GENE-0\" : [exp1[\"data\"][\"GENE-0\"][0],\"Dark Blue\"],\n",
    "        # \"GENE-1\" : [exp1[\"data\"][\"GENE-1\"][0],\"Dark Blue\"],\n",
    "        \"RE\" : [exp1[\"data\"][\"RE\"][0],\"Very Dark Gray\"],\n",
    "\n",
    "    },\n",
    "    \"config\":  exp1[\"config\"]\n",
    "}\n",
    "\n",
    "clear_output(wait=True)\n",
    "fig, ax = plt.subplots()\n",
    "means1b , stds =  plot_all(exp1b, ax)\n",
    "# plt.savefig('GENE-0_1e7_S_1000.png', dpi=600)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotting means\n",
    "\n",
    "\n",
    "steps = len(means1[0]) \n",
    "print(steps)\n",
    "thresholds = [0.938,0.939, 0.940]\n",
    "\n",
    "gm0 = []\n",
    "gm1 = []\n",
    "rr= []\n",
    "for t in thresholds:\n",
    "    for i in range(steps):\n",
    "        if means1[0][i] > t:\n",
    "            gm0.append(i)\n",
    "            break\n",
    "    for i in range(steps):\n",
    "        if means1[1][i] > t:\n",
    "            gm1.append(i)\n",
    "            break\n",
    "    for i in range(steps):\n",
    "        if means1[2][i] > t:\n",
    "            rr.append(i)\n",
    "            break\n",
    "print(gm0)\n",
    "print(rr)\n",
    "\n",
    "print(\"thr |  GENE-0 | GENE-1  |RE | speedup of gm over rr \")\n",
    "for i in range(len(thresholds)):\n",
    "    if gm1[i] <= rr[i]:\n",
    "        speed_up = \"+\" + str(int(100-(gm1[i]/rr[i]*100))) + \"%\"\n",
    "    else: \n",
    "        speed_up = str(int(100-(gm1[i]/rr[i]*100))) + \"%\"\n",
    "    print(f\"{thresholds[i]:.6} | {gm1[i]} | {gm1[i]}  |  {rr[i]} |  {speed_up} \")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "formatted_thresholds = [f\"{t:.4f}\" for t in thresholds[:-1]]\n",
    "formatted_thresholds.append(\"max\")\n",
    "speedups = []\n",
    "for i in range(len(thresholds)):\n",
    "    if gm1[i] <= rr[i]:\n",
    "        speedup = \"+\" + str(int(100 - (gm1[i] / rr[i] * 100))) + \"\\\\%\"\n",
    "    else:\n",
    "        speedup = str(int(100 - (gm1[i] / rr[i] * 100))) + \"\\\\%\"\n",
    "    speedups.append(speedup)\n",
    "    \n",
    "print(\"\\\\begin{tabular}{|c|\" + \"c|\" * len(formatted_thresholds) + \"}\")\n",
    "print(\"\\\\hline\")\n",
    "print(\"accuracy & \" + \" & \".join(formatted_thresholds) + \" \\\\\\\\\")\n",
    "print(\"\\\\hline\")\n",
    "print(\"GENE-0 & \" + \" & \".join([str(g) for g in gm0]) + \" \\\\\\\\\")\n",
    "print(\"\\\\hline\")\n",
    "print(\"GENE-1 & \" + \" & \".join([str(g) for g in gm1]) + \" \\\\\\\\\")\n",
    "print(\"\\\\hline\")\n",
    "print(\"RE & \" + \" & \".join([str(r) for r in rr]) + \" \\\\\\\\\")\n",
    "print(\"\\\\hline\")\n",
    "print(\"speedup & \" + \" & \".join(speedups) + \" \\\\\\\\\")\n",
    "print(\"\\\\hline\")\n",
    "print(\"\\\\end{tabular}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
